packages/suite/src/components/suite/modals/AddToken/index.tsx

```diff
@@ -0,0 +1,141 @@
+                coin: account.symbol,
+                descriptor: account.descriptor,
+                details: 'tokenBalances',
+                contractFilter: contractAddress,
+            });
+
+            if (response.success) {
+                const isInvalidToken = response.payload.tokens?.find(
+                    t => t.address === t.name && t.decimals === 0 && !t.symbol,
+                );
+                if (!isInvalidToken) {
+                    setTokenInfo(response.payload.tokens);
+                } else {
+                    // not a valid token
+                    setError(translationString('TR_ADD_TOKEN_TOKEN_NOT_VALID'));
+                }
+            } else {
+                setTokenInfo(undefined);
+                setError(
+                    translationString('TR_ADD_TOKEN_TOAST_ERROR', {
+                        error: response.payload.error,
+                    }),
+                );
+            }
+            setIsFetching(false);
+        };
+
+        loadTokenInfo();
+    }, [account, contractAddress, error, translationString]);
+
+    // it shouldn't be possible to open this modal without having selected account
+    if (!account) return null;
+
+    const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+        const addr = e.target.value;
+
+        const alreadyAdded = account.tokens?.find(
+            t => t.address.toLowerCase() === addr.toLowerCase(),
+        );
+
+        const isValid = isAddressValid(addr, account.symbol);
+
+        if (addr && !isValid) {
+            setError(translationString('TR_ADD_TOKEN_ADDRESS_NOT_VALID'));
+        } else if (alreadyAdded) {
+            setError(translationString('TR_ADD_TOKEN_ADDRESS_DUPLICATE'));
+        } else {
+            setError(null);
+        }
+        setTokenInfo(undefined);
+        setContractAddress(addr);
+    };
+
+    const getInputState = () => {
+        if (error) return 'error';
+        if (!error && contractAddress && !isFetching) return 'success';
+        return undefined;
+    };
+
+    return (
+        <Modal
+            cancelable
+            onCancel={props.onCancel}
+            heading={<Translation id="TR_ADD_TOKEN_TITLE" />}
+        >
+            <Wrapper>
+                <Input
+                    label={
+                        <QuestionTooltip
+                            label={<Translation id="TR_ADD_TOKEN_LABEL" />}
+                            tooltip="TR_ADD_TOKEN_TOOLTIP"
+                        />
+                    }
+                    placeholder={translationString('TR_ADD_TOKEN_PLACEHOLDER')}
+                    value={contractAddress}
+                    bottomText={error}
+                    state={getInputState()}
+                    onChange={onChange}
+                />
+            </Wrapper>
+
+            <Actions>
+                <Button
+                    onClick={() => {
+                        if (tokenInfo) {
```

Comment:
might not be needed as the button is disabled if there is no `tokenInfo`

Reply:
I guess author wanted to satisfy typescript here.

Reply:
@mroz22 is correct, can't call the function with undefined. typescript knows nothing about our UI constraints